File: packages/python/dt3_api/timer.py
"""Public timing lifecycle contract for the DT3 Python SDK."""

from __future__ import annotations

from typing import Protocol


class Timer(Protocol):
    """Measure an operation and emit its final duration through a logger."""

    # PUBLIC_INTERFACE
    def start(self) -> "Timer":
        """Start the timer and return it for fluent use."""

    # PUBLIC_INTERFACE
    def stop(self) -> float:
        """Stop the timer, emit its completion event, and return elapsed milliseconds."""

    # PUBLIC_INTERFACE
    def finish(self) -> float:
        """Alias for stop that emits the completion event and returns elapsed milliseconds."""

File: packages/node/src/api/Timer.ts
import { Context } from './types';

/**
 * Optional metadata attached to the timer completion event.
 */
export type TimerContext = Context;

/**
 * Measures one operation and emits its final duration through a logger.
 */
export interface Timer {
  /**
   * Start the timer and return it for fluent use.
   *
   * @returns This timer instance.
   * @throws Error when the timer has already started or its logger is closed.
   */
  start(): Timer;

  /**
   * Stop the timer, emit one completion event, and return elapsed milliseconds.
   *
   * @returns The non-negative elapsed duration in milliseconds.
   * @throws Error when the timer has not started, is already stopped, or its logger is closed.
   */
  stop(): number;

  /**
   * Alias for {@link stop}.
   *
   * @returns The non-negative elapsed duration in milliseconds.
   */
  finish(): number;
}
